presto-sqlserver/src/main/java/com/facebook/presto/plugin/sqlserver/SqlServerRecordCursor.java

```diff
@@ -0,0 +1,284 @@
+import java.sql.Statement;
+import java.sql.Time;
+import java.sql.Timestamp;
+import java.util.Calendar;
+import java.util.List;
+import java.util.TimeZone;
+import java.util.concurrent.TimeUnit;
+
+import static com.facebook.presto.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.DateTimeEncoding.packDateTimeWithZone;
+import static com.facebook.presto.spi.type.DateType.DATE;
+import static com.facebook.presto.spi.type.IntegerType.INTEGER;
+import static com.facebook.presto.spi.type.RealType.REAL;
+import static com.facebook.presto.spi.type.SmallintType.SMALLINT;
+import static com.facebook.presto.spi.type.TimeType.TIME;
+import static com.facebook.presto.spi.type.TimestampType.TIMESTAMP;
+import static com.facebook.presto.spi.type.TimestampWithTimeZoneType.TIMESTAMP_WITH_TIME_ZONE;
+import static com.facebook.presto.spi.type.TinyintType.TINYINT;
+import static com.facebook.presto.spi.type.VarbinaryType.VARBINARY;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static io.airlift.slice.Slices.utf8Slice;
+import static io.airlift.slice.Slices.wrappedBuffer;
+import static java.lang.Float.floatToRawIntBits;
+import static java.util.Objects.requireNonNull;
+
+class SqlServerRecordCursor
+        implements RecordCursor
+{
+    private static final Logger log = Logger.get(SqlServerRecordCursor.class);
+
+    private static final TimeZone UTC = TimeZone.getTimeZone("UTC");
+
+    private final List<JdbcColumnHandle> columnHandles;
+    private final Connection connection;
+    private final PreparedStatement statement;
+    private final ISQLServerResultSet resultSet;
+
+    private boolean closed;
+
+    public SqlServerRecordCursor(JdbcClient jdbcClient, JdbcSplit split, List<JdbcColumnHandle> columnHandles)
+    {
+        this.columnHandles = ImmutableList.copyOf(requireNonNull(columnHandles, "columnHandles is null"));
+        try {
+            this.connection = requireNonNull(jdbcClient, "jdbcClient is null").getConnection(split);
+            this.statement = jdbcClient.buildSql(this.connection, split, columnHandles);
+            log.debug("Executing: %s", statement.toString());
+            this.resultSet = this.statement.executeQuery().unwrap(ISQLServerResultSet.class);
+        }
+        catch (SQLException | RuntimeException e) {
+            throw this.handleSqlException(e);
+        }
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return 0;
+    }
+
+    @Override
+    public long getTotalBytes()
+    {
+        return 0;
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return 0;
+    }
+
+    @Override
+    public Type getType(int field)
+    {
+        return this.columnHandles.get(field).getColumnType();
+    }
+
+    @Override
+    public boolean advanceNextPosition()
```

Comment:
The constructor is public. Are you talking about `handleSqlException`? It's fine to make that method protected. We'd rather not duplicate a lot of code just to avoid changing visibility.